gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+import org.hypertrace.gateway.service.v1.baseline.BaselineEntity;
+import org.hypertrace.gateway.service.v1.baseline.BaselineEntitiesRequest;
+import org.hypertrace.gateway.service.v1.baseline.BaselineEntitiesResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This service calculates baseline values for an Entity across time range. This converts Functions
+ * in selection expression to time series data and generate baseline, lower bound and upper bound.
+ * It also changes Time aggregation requests to get more data points for proper baseline calculation
+ */
+public class BaselineServiceImpl implements BaselineService {
+  private static long DAY_IN_MILLIS = 86400000L;
+  private static long SIX_HOURS_IN_MILLIS = 21600000L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(BaselineServiceImpl.class);
+  private final AttributeMetadataProvider attributeMetadataProvider;
+  private final BaselineServiceQueryParser baselineServiceQueryParser;
+  private final BaselineServiceQueryExecutor baselineServiceQueryExecutor;
+
+  public BaselineServiceImpl(
+      AttributeMetadataProvider attributeMetadataProvider,
+      BaselineServiceQueryParser baselineServiceQueryParser,
+      BaselineServiceQueryExecutor baselineServiceQueryExecutor) {
+    this.attributeMetadataProvider = attributeMetadataProvider;
+    this.baselineServiceQueryParser = baselineServiceQueryParser;
+    this.baselineServiceQueryExecutor = baselineServiceQueryExecutor;
+  }
+
+  public BaselineEntitiesResponse getBaselineForEntities(
+      String tenantId,
+      BaselineEntitiesRequest originalRequest,
+      Map<String, String> requestHeaders) {
+    BaselineRequestContext requestContext =
+        getRequestContext(tenantId, requestHeaders, originalRequest);
+    String timeColumn =
+        AttributeMetadataUtil.getTimestampAttributeId(
+            attributeMetadataProvider, requestContext, originalRequest.getEntityType());
+    Map<String, BaselineEntity> baselineEntityAggregatedMetricsMap = Collections.EMPTY_MAP;
+    Map<String, BaselineEntity> baselineEntityTimeSeriesMap = Collections.EMPTY_MAP;
+
+    if (originalRequest.getBaselineAggregateRequestCount() > 0) {
+      // Aggregated Functions data
+      Period aggTimePeriod =
+          getPeriod(originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      long periodSecs = getPeriodInSecs(aggTimePeriod);
+      List<TimeAggregation> timeAggregations = getTimeAggregationsForAggregateExpr(originalRequest);
+      updateAliasMap(requestContext, timeAggregations);
+      // Take more data to calculate baseline and standard deviation.
+      long seriesStartTime =
```

Comment:
This is the issue from the last round of comments that I'm most concerned about. It's sourcing the data incorrectly as a baseline shouldn't use data from the time range it's applicable for, and worse - the later implementation implicitly relies on that behavior (see my other comment).

Reply:
Suppose user selected last 6 hours of data and the traffic increased from last 4 hours in user environment, if we don't include the current data for baseline won't the values deviate too much from baseline ?

Reply:
This is precisely the use case of baselines as I understand them, and why I've really been pushing for this change - seeing that the data in the requested range is higher than it historically has been or could be predicted. If the calculation includes the current data, then as you pointed out, it would be more difficult to identify deviations.

Reply:
ok, will update this.